benchmark/src/runner.rs

```diff
@@ -0,0 +1,321 @@
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+use anyhow::Result;
+use clippy_utilities::{NumericCast, OverflowArithmetic};
+use indicatif::ProgressBar;
+use log::info;
+use rand::RngCore;
+use tokio::{
+    sync::{
+        mpsc::{self, Receiver},
+        Barrier,
+    },
+    time::{Duration, Instant},
+};
+use xline::client::{kv_types::PutRequest, Client};
+
+use crate::{args::Commands, Benchmark};
+
+/// Result of request
+#[derive(Debug)]
+struct CmdResult {
+    /// elapsed time
+    elapsed: Duration,
+    /// Error message of request
+    error: Option<String>,
+}
+
+/// `CommandRunner` is the main struct for running commands.
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug)]
+pub struct CommandRunner {
+    /// errors of requests
+    errors: HashMap<String, usize>,
+    /// Args of benchmark
+    args: Benchmark,
+}
+
+#[derive(Debug, Default)]
+
+/// Stats of benchmark
+pub struct Stats {
+    /// all latencies
+    pub latencies: Vec<Duration>,
+    /// total time
+    pub total: Duration,
+    /// slowest latency
+    pub slowest: Duration,
+    /// fastest latency
+    pub fastest: Duration,
+    /// average latency
+    pub avg: Duration,
+    /// Operations per second
+    pub qps: f64,
+}
+
+impl Stats {
+    /// New `Stats`
+    fn new() -> Self {
+        Self::default()
+    }
+
+    /// Summary of stats
+    pub fn summary(&self) -> String {
+        let mut s = String::from("Summary:\n");
+        s.push_str(&format!(
+            "  Total:        {:.4} secs\n",
+            self.total.as_secs_f64()
+        ));
+        s.push_str(&format!(
+            "  Slowest:      {:.4} secs\n",
+            self.slowest.as_secs_f64()
+        ));
+        s.push_str(&format!(
+            "  Fastest:      {:.4} secs\n",
+            self.fastest.as_secs_f64()
+        ));
+        s.push_str(&format!(
+            "  Average:      {:.4} secs\n",
+            self.avg.as_secs_f64()
+        ));
+        s.push_str(&format!("  Requests/sec: {:.2}\n", self.qps));
+        s
+    }
+
+    /// Calculate the histogram from the latencies.
+    #[allow(clippy::indexing_slicing)]
+    pub fn histogram(&self) -> String {
```

Comment:
Is there any histogram lib existed? If not, can we provide one?

Reply:
I found a lib but can't output a graph like this